File: npc_app/national_petro_carbon/document_controllers/delivery_note.py
import frappe

def auto_fill_from_picklist(doc, method):
    """
    Auto-fill transporter and vehicle_no in Delivery Note
    from Pick List linked to this DN.
    """
    # Loop through items to find Pick List
    for item in doc.items:
        if item.against_pick_list:  # standard link to Pick List
            pick_list_doc = frappe.get_doc("Pick List", item.against_pick_list)

            # Map fields from Pick List to DN
            doc.transporter_name= pick_list_doc.custom_transporter_name_
            doc.vehicle_no = pick_list_doc.custom_vehicle_no

            # Only take first Pick List
            break

File: npc_app/public/js/delivery_note.js
frappe.ui.form.on("Delivery Note", {
    onload(frm) {
        // Only run when DN is created from Pick List
        if (!frm.doc.pick_list) return;

        frappe.db.get_value(
            "Pick List",
            frm.doc.pick_list,
            ["custom_transporter_name_", "custom_vehicle_no"],
            (r) => {
                if (!r) return;

                // Set Transporter
                if (r.custom_transporter_name_ && !frm.doc.transporter) {
                    frm.set_value("transporter", r.custom_transporter_name_);
                }

                // Set Vehicle No (More Info section)
                if (r.custom_vehicle_no && !frm.doc.vehicle_no) {
                    frm.set_value("vehicle_no", r.custom_vehicle_no);
                }

                frm.refresh_fields(["transporter", "vehicle_no"]);
            }
        );
    }
});
